dali/imgcodec/image_decoder.cc

```diff
@@ -351,45 +351,33 @@ void ImageDecoder::DecoderWorker::process_batch(std::unique_ptr<ScheduledWork> w
                                  work->params,
                                  make_span(work->rois));
 
-    if (fallback_) {
-      for (;;) {
-        auto indices = future.wait_new();
-        if (indices.empty())
-          break;
-
-        for (int sub_idx : indices) {
-          DecodeResult r = future.get_one(sub_idx);
-          if (r.success) {
-            if (!decode_to_gpu && !work->gpu_outputs.empty()) {
-              copy(work->gpu_outputs[sub_idx], work->cpu_outputs[sub_idx], work->ctx.stream);
-            }
-            work->results.set(work->indices[sub_idx], r);
-          } else {
+    for (;;) {
+      auto indices = future.wait_new();
+      if (indices.empty())
+        break;  // if wait_new returns with an empty result, it means that everything is ready
+
+      for (int sub_idx : indices) {
+        DecodeResult r = future.get_one(sub_idx);
+        if (r.success) {
+          if (!decode_to_gpu && !work->gpu_outputs.empty()) {
+            copy(work->gpu_outputs[sub_idx], work->cpu_outputs[sub_idx], work->ctx.stream);
+          }
+          work->results.set(work->indices[sub_idx], r);
+        } else {  // failed to decode
+          if (fallback_) {
+            // if there's fallback, we don't set the result, but try to use the fallback first
             if (!fallback_work)
               fallback_work = owner_->new_work(work->ctx, work->results, work->params);
             fallback_work->move_entry(*work, sub_idx);
+          } else {
+            // no fallback - just propagate the result to the original promise
+            work->results.set(work->indices[sub_idx], r);
           }
         }
-
-        if (fallback_work && !fallback_work->empty())
-          fallback_->add_work(std::move(fallback_work));
-      }
-    } else {
-      // no fallback - the results are final
-      future.wait_all();
```

Comment:
I thought it would be better to wait for all results and submit the H2D copies in one go - I was wrong.